visualization/utils/common.py

```diff
@@ -0,0 +1,388 @@
+                    type_function = int
+                elif str(value).replace(".", "", 1).replace("-", "", 1).isdigit():
+                    type_function = float
+
+                values[input_group_child["props"]["id"]] = type_function(str(value).replace('\\n', '\n'))
+
+    return values
+
+
+def get_stats(all_files_data: List[Dict], is_correct: bool = True) -> float:
+    right = 0
+    for data in all_files_data:
+        right += data.get("is_correct", not is_correct) == is_correct
+
+    return right / len(all_files_data) if len(all_files_data) else -1
+
+
+def get_metrics(all_files_data: List[Dict], errors_dict: Dict = {}) -> Dict:
+    correct_responses = get_stats(all_files_data, True)
+    wrong_responses = get_stats(all_files_data, False)
+    no_response = 1.0 - correct_responses - wrong_responses
+    custom_stats = {}
+    for name, func in get_custom_stats().items():
+        if name not in errors_dict:
+            errors_dict[name] = {}
+        custom_stats[name] = catch_eval_exception(
+            [],
+            func,
+            all_files_data,
+            "Got error when applying function",
+            errors_dict[name],
+        )
+
+    stats = {
+        'correct_responses': round(correct_responses, 2),
+        "wrong_responses": round(wrong_responses, 2),
+        "no_response": round(no_response, 2),
+        **custom_stats,
+    }
+    return stats
+
+
+def get_eval_function(text):
+    template = """
+def eval_function(data):
+{}
+    return {}
+"""
+    code_lines = [''] + text.strip().split('\n')
+    code = template.format(
+        '\n    '.join(code_lines[:-1]),
+        code_lines[-1:][0],
+    )
+    namespace = {}
+    exec(code, namespace)
+    return namespace['eval_function']
+
+
+def calculate_metrics_for_whole_data(table_data: List, model_id: str) -> Dict:
+    errors_dict = {}
+    for question_id in range(len(table_data)):
+        stats = get_metrics(table_data[question_id][model_id], errors_dict)
+        table_data[question_id][model_id] = list(
+            map(
+                lambda data: {**data, **stats},
+                table_data[question_id][model_id],
+            )
+        )
+    if len(errors_dict):
+        for name, error_dict in errors_dict.items():
+            logging.error(ERROR_MESSAGE_TEMPLATE.format(name, error_dict))
+
+
+def catch_eval_exception(
+    available_models: List[str],
+    eval_func: Callable[[Dict], bool],
+    data: Dict,
+    default_answer: Union[bool, str],
+    errors_dict: Optional[Dict] = {},
+) -> bool:
+    try:
+        if eval_func is None:
+            return default_answer
+        return eval_func(data)
+    except Exception as e:
+        if str(e).split(" ")[-1].replace("'", "") not in available_models:
+            if str(e) not in errors_dict:
+                errors_dict[str(e)] = 0
+            errors_dict[str(e)] += 1
+        return default_answer
+
+
+def custom_deepcopy(data) -> List:

```

Comment:
why is this needed?

Reply:
1. it works faster
2. it does not do deepcopy for items in last json. So when I update them they are updated everywhere